transport-native-epoll/src/main/c/io_netty_channel_libaio_LibaioContext.c

```diff
@@ -0,0 +1,390 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+
+#ifndef _GNU_SOURCE
+// libaio, O_DIRECT and other things won't be available without this define
+#define _GNU_SOURCE
+#endif
+
+//#define DEBUG
+
+#include <jni.h>
+#include <unistd.h>
+#include <errno.h>
+#include <libaio.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <stdlib.h>
+#include <pthread.h>
+#include "io_netty_channel_libaio_LibaioContext.h"
+#include "exception_helper.h"
+
+struct io_control {
+    io_context_t ioContext;
+    struct io_event * events;
+
+    // This is used to make sure we don't return IOCB while something else is using them
+    // this is to guarantee the submits could be done concurrently with polling
+    pthread_mutex_t iocbLock;
+
+    // a resuable pool of iocb
+    struct iocb ** iocb;
+    int queueSize;
+    int iocbPut;
+    int iocbGet;
+    int used;
+};
+
+jclass errorInfoClass = NULL;
+jmethodID errorInfoConstr = NULL;
+
+/**
+  This is being invoked by Native.c /  JNI_OnLoad..
+
+  There is only one point of entrance of Loading the library,
+  this will be a hook so extra stuff that needs to be loaded here.
+*/
+jint directFile_JNI_OnLoad(JNIEnv* env) {
+    errorInfoClass = (*env)->FindClass(env, "io/netty/channel/libaio/ErrorInfo");
+    if (errorInfoClass == NULL) {
+       return JNI_ERR;
+    }
+
+    // The ErrorInfoClass is barely used. The VM would crash in the event of an error without this GlobalRef
+    // and if the class was unloaded from GC
+    errorInfoClass = (jclass)(*env)->NewGlobalRef(env, (jobject)errorInfoClass);
+
+    errorInfoConstr = (*env)->GetMethodID(env, errorInfoClass, "<init>", "(Ljava/lang/Object;ILjava/lang/String;)V");
+
+    if (errorInfoConstr == NULL) {
+       return JNI_ERR;
+    }
+
+    /// The ErrorInfoClass is barely used. The VM would crash in the event of an error without this GlobalRef
+    errorInfoConstr = (jmethodID)(*env)->NewGlobalRef(env, (jobject)(errorInfoConstr));
+
+    return JNI_VERSION_1_6;
+}
+
+void directFile_JNI_OnUnLoad(JNIEnv* env) {
+    // Deleting global refs so their classes can be GCed
+    if (errorInfoConstr != NULL) {
+        (*env)->DeleteGlobalRef(env, (jobject)errorInfoConstr);
+    }
+
+    if (errorInfoClass != NULL) {
+        (*env)->DeleteGlobalRef(env, (jobject)errorInfoClass);
+    }
+}
+
+static inline struct io_control * getIOControl(JNIEnv* env, jobject pointer) {
+    struct io_control * ioControl = (struct io_control *) (*env)->GetDirectBufferAddress(env, pointer);
+    return ioControl;
+}
+
+/**
+ remove an iocb from the pool of IOCBs. Returns null if full
```

Comment:
I didn't realize you were using javadoc style comments on Native.c

i will fix all the other ones